### DISCUSSION TITLE
RedwoodJS on Vercel

### DISCUSSION THREAD

**Discussion by aengl** [vercel/vercel#4535] [Help] [answered]
Hello everyone, I'm migrating a webapp to https://redwoodjs.com/ but would like to avoid having to migrate the hosting to Netlify. Anyone else interested in getting RedwoodJS to run on Vercel?

So far, my naive approach is to add the following `vercel.json`:

```
{
  "version": 2,
  "routes": [
    { "handle": "filesystem" },
    { "src": "/.*", "dest": "/web/dist/index.html" }
  ]
}
```

In `Build & Development Settings`, I choose `Other` as the preset and change the output directory to `web/dist/` and the build command to `yarn rw db up --no-db-client && yarn rw build`.

My `redwood.toml` looks as follow:

```
[web]
  port = 8910
  apiProxyPath = "/api/dist/functions"
[api]
  port = 8910
[browser]
  open = true
```

Other than Vercel building way more functions than necessary (can't find anything in the documentation on restricting the functions to `api/dist`), so far so good.

Unfortunately, all requests fail with 502, with the log file saying:
```
[POST] /api/dist/functions/graphql
ERROR	Error: @prisma/client did not initialize yet. Please run "prisma generate" and try to import it again.
```

Ideas are welcome!

**Comment by aengl** [comment]
Relevant discussion on how to make Prisma run on Vercel: https://github.com/prisma/prisma-client-js/issues/390

**Comment by styfle** [comment]
It looks like RedwoodJS uses the AWS Lambda API which can be enabled with an environment variable.

https://vercel.com/docs/runtimes#advanced-usage/advanced-node-js-usage/aws-lambda-api

We are planning to add zero config support for RedwoodJS so I'll post back here when that is ready.

**Comment by styfle** [comment]
@aengl I see the same `@prisma/client did not initialize yet` error with Netlify when I used the `yarn create redwood-app` boilerplate. 
It seems like this is likely a bug with Redwood or Prisma. 

I also see that Redwood has a note in the readme:

> **WARNING**: Redwood is still in the alpha phase of development. We do not recommend that you use Redwood applications in production at this time. 

>**HEADS UP:** RedwoodJS is _NOT_ ready for use in Production. It relies heavily on Prisma2, which is currently in testing with an expected production release coming soon. See status at ["Is Prisma2 Ready?"](https://isprisma2ready.com)

**Comment by styfle** [comment]
Also, its not clear what `apiProxyPath` is supposed to do.

The documentation says "TODO"

https://redwoodjs.com/docs/app-configuration-redwood-toml.html#webapiproxypaths

![image](https://user-images.githubusercontent.com/229881/84834402-40fc0180-afff-11ea-8789-26759b80f7f4.png)

**Comment by aengl** [comment]
Yes, I will open an issue with them regarding that.

It also seems to be misnamed. They bootstrap new projects with the following configuration:

```
[web]
  port = 8910
  apiProxyPath = "/.netlify/functions"
[api]
  port = 8911
[browser]
  open = true
```

Note that it's `apiProxyPath`, not `apiProxyPaths`. The configuration results in the lambda requests going to `/.netlify/functions/graphql`. With Vercel, that would probably be `/api`, or `/api/dist/functions` if the output path is preserved.

The disclaimer for Redwood is a bit outdated, Prisma 2 is no longer in beta and production ready. I don't know what that means for the production readiness of RedwoodJS, but I think the problem is essentially about how to host Prisma 2 on a lambda on Vercel, which I think is of interest beyond users of RedwoodJS 😃

**Comment by styfle** [comment]
Do the serverless functions run locally with `yarn rw dev`? It looks like they run on a completely separate port (8911 vs the expected 8910) but that is probably not how it works in production deployments, right?

I tried `http://localhost:8910/api/graphql`, `http://localhost:8910/api/functions/graphql`, and `http://localhost:8910/graphql` to execute the function but I just see the 404 page.

**Reply by aengl** [reply]
Yes, locally they're hosted via 8911, but it's the default HTTP port on production. That's what the `redwood.toml` does with
```
[api]
  port = 8911
```

I think all those port configurations are ignored on production, it just seems to be port 80 for everything.

**Comment by styfle** [comment, marked as answer]
Hi @aengl 

You can now use RedwoodJS without `vercel.json`. When you import a project, RedwoodJS will be autodetected.

Alternatively, you can change an existing project to use redwood in the project's general settings.

If you plan to use Vercel CLI, make sure you use the latest [20.0.0 release](https://github.com/vercel/vercel/releases/tag/vercel%4020.0.0).

**Comment by aengl** [comment]
That is amazing news, thank you! I'll migrate my app soon and will let you know how it goes.

**Comment by aengl** [comment]
Zero-config deployment worked like a charm, thank you once more for your incredible work and support! 💛
